### PR TITLE
don't fail on first http error in STAC provider

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
[Retry STAC Request](https://trello.com/c/d91PY3eu/415-retry-stac-request)

File: services/src/pro/datasets/external/sentinel_s2_l2a_cogs.rs [hunk 1]
```diff
@@ -360,23 +362,30 @@ impl SentinelS2L2aCogsMetaData {
...
-            .send()
-            .await
-            .context(error::Reqwest)?
-            .text()
-            .await
-            .context(error::Reqwest)?;
+        let client = Client::builder().build()?;
+
+        // TODO: configure max retries and delay, also find good defaults
+        retry(3, 125, 2., async || -> Result<StacCollection> {
```
Comment (michaelmattig) [inline services/src/pro/datasets/external/sentinel_s2_l2a_cogs.rs:368]: I think the magic numbers should be config params. Either for the provider or globally.

File: services/src/util/retry.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+pub async fn retry<F, T, E, Fut>(
+    max_retries: usize,
+    initial_delay_ms: u64,
+    exponential_backoff_factor: f64,
+    mut f: F,
+) -> Result<T, E>
+where
+    F: FnMut() -> Fut,
+    Fut: Future<Output = Result<T, E>>,
+    T: std::fmt::Debug, // TODO: remove
```
Comment (michaelmattig) [inline services/src/util/retry.rs:23]: remove

File: services/src/util/retry.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    E: std::fmt::Debug, // TODO: remove
```
Comment (michaelmattig) [inline services/src/util/retry.rs:24]: remove

File: services/src/util/retry.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+/// A method wrapper for calling a method that may fail spuriously until it succeeds.
+/// The method is called at most `max_retries` times.
+/// If it still fails after `max_retries` times, the error is returned.
+///
+/// Uses exponential backoff by taking the `initial_delay_ms` and multiplying each time an `exponential_backoff_factor` on it.
+///
+/// # Panics
+/// Panics if `max_retries` is 0.
+///
+pub async fn retry<F, T, E, Fut>(
```
Comment (michaelmattig) [inline services/src/util/retry.rs:14]: can/should we have a test just for this method?

File: test_data/pro/stac_responses/items_page_1_limit_500.json [hunk 5]
```diff
@@ -0,0 +1,390 @@
+{
```
Comment (michaelmattig) [inline test_data/pro/stac_responses/items_page_1_limit_500.json:1]: are we allowed to store this file in our repo license wise?

**Review by michaelmattig — CHANGES_REQUESTED** [review]

File: test_data/pro/stac_responses/items_page_1_limit_500.json (hunk 5 above)
Comment (ChristianBeilschmidt) [reply test_data/pro/stac_responses/items_page_1_limit_500.json:1]: Hm, the STAC API is under MIT https://github.com/stac-utils/stac-server
Maybe we need to remove the URLs?
But there is not much Schöpfungshöhe.

File: services/src/pro/datasets/external/sentinel_s2_l2a_cogs.rs (hunk 1 above)
Comment (ChristianBeilschmidt) [reply services/src/pro/datasets/external/sentinel_s2_l2a_cogs.rs]: a8c60bf

File: services/src/util/retry.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply services/src/util/retry.rs:14]: 9c9ef30

File: services/src/util/retry.rs (hunk 3 above)
Comment (ChristianBeilschmidt) [reply services/src/util/retry.rs]: 9c9ef30

File: services/src/util/retry.rs (hunk 2 above)
Comment (ChristianBeilschmidt) [reply services/src/util/retry.rs]: 9c9ef30

File: services/src/util/retry.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    exponential_backoff_factor: f64,
+    mut f: F,
+) -> Result<T, E>
+where
+    F: FnMut() -> Fut,
+    Fut: Future<Output = Result<T, E>>,
+    T: std::fmt::Debug, // TODO: remove
+    E: std::fmt::Debug, // TODO: remove
+{
+    assert!(max_retries > 0, "Must use at least one retry");
```
[the remark this answers is not in the document]
Comment (ChristianBeilschmidt) [reply services/src/util/retry.rs]: 9c9ef30

**Review by michaelmattig — APPROVED** [review]

**Pull request merged**